consensus/istanbul/backend/announce.go

```diff
@@ -17,208 +17,386 @@
+			switch announceGossipFrequencyState {
+			case HighFreqBeforeFirstPeerState:
+				if len(sb.broadcaster.FindPeers(nil, p2p.AnyPurpose)) > 0 {
+					announceGossipFrequencyState = HighFreqAfterFirstPeerState
+				}
+
+			case HighFreqAfterFirstPeerState:
+				if numGossipedMsgsInHighFreqAfterFirstPeerState >= 10 {
+					announceGossipFrequencyState = LowFreqState
+				}
+				numGossipedMsgsInHighFreqAfterFirstPeerState += 1
+
+			case LowFreqState:
+				if currentAnnounceGossipTickerDuration != time.Duration(sb.config.AnnounceGossipPeriod)*time.Second {
+					// Reset the ticker
+					currentAnnounceGossipTickerDuration = time.Duration(sb.config.AnnounceGossipPeriod) * time.Second
+					announceGossipTicker.Stop()
+					announceGossipTicker = time.NewTicker(currentAnnounceGossipTickerDuration)
+					announceGossipTickerCh = announceGossipTicker.C
+				}
+			}
+
+			go sb.generateAndGossipAnnounce()
+
+			// Use this timer to also prune all announce related data structures.
+			if err := sb.pruneAnnounceDataStructures(); err != nil {
+				logger.Warn("Error is pruning announce data structures", "err", err)
+			}
+
+		case <-announceVersionsCheckTicker.C:
+			logger.Trace("Going to check peers' announce version set")
+			go sb.checkPeersAnnounceVersions()
+
+		case <-sb.announceThreadQuit:
+			checkIfShouldAnnounceTicker.Stop()
+			announceVersionsCheckTicker.Stop()
+			if announceGossipTicker != nil {
+				announceGossipTicker.Stop()
+			}
+			return
+		}
+	}
 }
 
-type announceData struct {
-	AnnounceRecords []*announceRecord
-	EnodeURLHash    common.Hash
-	Timestamp       uint
+func (sb *Backend) shouldGenerateAndProcessAnnounce() (bool, error) {
+
+	// Check if this node is in the registered/elected validator set
+	regAndElectedVals, err := sb.retrieveRegisteredAndElectedValidators()
+	if err != nil {
+		return false, err
+	}
+
+	return sb.coreStarted && regAndElectedVals[sb.Address()], nil
 }
 
-func (ad *announceData) String() string {
-	return fmt.Sprintf("{Timestamp: %v, EnodeURLHash: %v, AnnounceRecords: %v}", ad.Timestamp, ad.EnodeURLHash.Hex(), ad.AnnounceRecords)
+// pruneAnnounceDataStructures will remove entries that are not in the registered/elected validator set from all announce related data structures.
+// The data structures that it prunes are:
+// 1)  cachedAnnounceMsgs
+// 2)  lastAnnounceGossiped
+// 3)  valEnodeTable
+func (sb *Backend) pruneAnnounceDataStructures() error {
+	logger := sb.logger.New("func", "pruneAnnounceDataStructures")
+
+	// retrieve the registered/elected validator set
+	regAndElectedVals, err := sb.retrieveRegisteredAndElectedValidators()
+	if err != nil {
+		return err
+	}
+
+	// Prune the announce cache for entries that are not in the current registered/elected validator set
+	for cachedValAddress := range sb.cachedAnnounceMsgs {
```

Comment:
again given we can be modifying `cachedAnnounceMsgs` in separate goroutines we should lock